### ISSUE TITLE
Case Class References Missing Instantiations

### ISSUE THREAD

**Issue by wchau** [sourcegraph/scip-java#403]
`lsif-java` version: `0.7.6`

Type of code: case classes in Scala

Expected behavior: Case class "Find Reference" should contain instantiations.

Actual behavior: Case class "Find References" does not contain instantiations. You can find other instantiations if you "Find Reference" any one instantiation.

If the companion object is explicitly defined, you can find instantiations there by "Find Reference".

Note that "Go to Definition" from instantiations does go to the case class if the companion object does not exist (otherwise, it would go to the companion object), but the hover references the companion object(???). Ideally, all the synthetic functions should reference the case class (and vice versa) similar to IntelliJ's code reference.

```
case class TestA(a: String) {}  // "Find References" missing the below instantation

...

TestA("blah")  // Hover references object TestA but "Go to Definition" jumps to case class definition
```

**Comment by wchau** [comment]
It is already a lot better than before, though, thanks! :)

**Comment by olafurpg** [comment]
Thank you for reporting! I am able to reproduce and I opened a PR fixing this issue in #404 

Please keep the bug reports coming, I appreciate your patience with all of the different releases. It's tricky to get this right for case classes since they generate so many different kinds of symbols (some which we want to include with "find ref" results while not others)

**Comment by olafurpg** [comment]
Reopening this issue since "Find references" on the case class does not show usages of the synthetic companion object. Here's a minimized test case that reproduces the problem

https://github.com/sourcegraph/scip-java/blob/bf4ca0c41b17b2866c5fc4fd8032c0e5eb0ccd71/tests/snapshots/src/main/generated/tests/minimized-scala/src/main/scala/minimized/Issue396.scala#L16

To fix this problem, we should add an `is_reference` relationship alongside the `is_definition` relationship. I estimate this is an easy issue to fix and it does not require changes in the Sourcegraph backend.

**Comment by olafurpg** [comment]
The root cause of this issue is here where we intentionally return `false` for objects https://github.com/sourcegraph/scip-java/blob/bf4ca0c41b17b2866c5fc4fd8032c0e5eb0ccd71/scip-semanticdb/src/main/java/com/sourcegraph/scip_semanticdb/ScipSemanticdb.java#L350

I'm wondering if the fix is to remove the `supportsReferenceRelationship` condition where we set `is_definition`. As long as the names of the symbols are the same then it's OK for them to share a reference relationship.

 https://github.com/sourcegraph/scip-java/blob/bf4ca0c41b17b2866c5fc4fd8032c0e5eb0ccd71/scip-semanticdb/src/main/java/com/sourcegraph/scip_semanticdb/ScipSemanticdb.java#L147

**Comment by olafurpg** [comment]
Another possible fix, which I think might be a better solution, is to add an `is_reference` relationship on the synthetic `apply().` on the companion object. This solution is better because it works the same regardless if there is an explicit companion or not.

**Comment by olafurpg** [comment]
Concretely, the proposal is to update this line here in the snapshots to include `is_reference`

https://github.com/sourcegraph/scip-java/blob/8964623f8a1a81e59b4f4cbc37951e61c828a535/tests/snapshots/src/main/generated/tests/minimized-scala/src/main/scala/minimized/Issue403.scala#L13

**Comment by olafurpg** [comment]
Fixed in #561 

Link to reproduce https://sourcegraph.com/github.com/sourcegraph/scip-java/-/blob/tests/minimized-scala/src/main/scala/minimized/Issue403.scala?L3:12-3:20#tab=references

![CleanShot 2023-04-24 at 18 14 33@2x](https://user-images.githubusercontent.com/1408093/234055668-3016f48d-f5fb-419b-8302-e26ddc6b75e0.png)

Note that the empty line is refers to a zero-length range for a usage of the `apply()` symbol. I have opened a separate issue to track that https://github.com/sourcegraph/scip-java/issues/564
